### PR TITLE
fix!: allow safe leverage operations during partial oracle outages

### PR DESCRIPTION
**Pull request by toteki**
## Description

This one does a lot of things:
- Allows `account_summary` to work during price outages, treating all unknown prices as zero.
  - In such cases, supplied value and other fields will appear lower than it really is, since some assets were skipped
  - Liquidation threshold will be null when it can't be computed, since there's no safe way to do that with missing prices
- `borrow_limit` in queries as well as messages is computed using only collateral with known prices
  - If the portion of your collateral with known prices is enough to cover a borrow, then it still works.
  - Same for withdraw and decollateralize
- `MaxWithdraw` and `MaxBorrow` (both queries and messages) now function with missing collateral prices, respecting the borrow limit policy above. The queries return zero on missing borrow prices, or a missing price for the specific token being asked for.
- `liquidation_targets` query skips addresses where liquidation threshold cannot be computed, instead of returning an error for the whole query
- `MsgLiquidate` will be able to function with some missing prices on the target's borrowed assets, if their other borrows are  high enough to still put them above their liquidation threshold

API Breaking:
- `liquidation_threshold` field in `account_summary` field can now be null

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/keeper/msg_server.go [hunk 1]
```diff
@@ -330,6 +342,10 @@ func (s msgServer) MaxBorrow(
 		return nil, err
 	}
 
+	// If a price is missing for the borrower's collateral,
+	// but not this uToken or any of their borrows, error
```
Comment (toteki) [inline x/leverage/keeper/msg_server.go:346]:
```suggestion
	// but not this token or any of their borrows, error
```

File: x/leverage/keeper/msg_server.go [hunk 2]
```diff
@@ -330,6 +342,10 @@ func (s msgServer) MaxBorrow(
 		return nil, err
 	}
 
+	// If a price is missing for the borrower's collateral,
+	// but not this token or any of their borrows, error
+	// will be nil and the resulting value will be what
+	// can safely be withdrawn even with missing prices.
```
Comment (toteki) [inline x/leverage/keeper/msg_server.go:348]:
```suggestion
	// can safely be borrowed even with missing prices.
```

File: x/leverage/keeper/borrows.go [hunk 3]
```diff
@@ -127,6 +130,43 @@ func (k Keeper) CalculateBorrowLimit(ctx sdk.Context, collateral sdk.Coins) (sdk
...
+		// ignore blacklisted tokens
+		if !ts.Blacklist {
+			// get USD value of base assets using the chosen price mode
+			v, err := k.TokenValue(ctx, baseAsset, types.PriceModeLow)
+			if err == nil {
+				// if both spot and historic (if requeired) prices exist,
+				// add each collateral coin's weighted value to borrow limit
+				limit = limit.Add(v.Mul(ts.CollateralWeight))
+			}
+
```
Comment (robert-zaremba) [inline x/leverage/keeper/borrows.go:163]: let's add `else` and check `err`, in some cases we should return an error, rather than ignoring.

Comment (robert-zaremba) [reply x/leverage/keeper/borrows.go:163]: in other words, let's ignore if an error `Is` of a specific type.

File: x/leverage/keeper/grpc_query.go [hunk 4]
```diff
@@ -320,11 +323,13 @@ func (q Querier) MaxWithdraw(
...
+		// but not this uToken or any of their borrows, error
+		// will be nil and the resulting value will be what
+		// can safely be withdrawn even with missing prices.
+		// On non-nil error here, max withdraw is zero.
 		uToken, err := q.Keeper.maxWithdraw(ctx, addr, denom)
-		if err != nil {
-			return nil, err
-		}
-		if uToken.IsPositive() {
+		if err == nil && uToken.IsPositive() {
```
Comment (robert-zaremba) [inline x/leverage/keeper/grpc_query.go:332]: if the `err` is not nil, then we should return or check it at least.

File: x/leverage/keeper/grpc_query.go [hunk 5]
```diff
@@ -374,14 +379,15 @@ func (q Querier) MaxBorrow(
...
+		// but not this token or any of their borrows, error
+		// will be nil and the resulting value will be what
+		// can safely be borrowed even with missing prices.
+		// On non-nil error here, max borrow is zero.
 		maxBorrow, err := q.Keeper.maxBorrow(ctx, addr, denom)
-		if err != nil {
-			return nil, err
-		}
-		if maxBorrow.IsPositive() {
+		if err == nil && maxBorrow.IsPositive() {
```
Comment (robert-zaremba) [inline x/leverage/keeper/grpc_query.go:388]: ditto

File: x/leverage/keeper/iter.go [hunk 6]
```diff
@@ -168,20 +168,16 @@ func (k Keeper) GetEligibleLiquidationTargets(ctx sdk.Context) ([]sdk.AccAddress
...
+		// to the next borrower instead of stopping the entire query
 		liquidationLimit, err := k.CalculateLiquidationThreshold(ctx, collateral)
-		if err != nil {
-			return err
-		}
-
-		// If liquidation limit is smaller than borrowed value then the
-		// address is eligible for liquidation.
-		if liquidationLimit.LT(borrowValue) {
+		if err == nil && liquidationLimit.LT(borrowValue) {
```
Comment (robert-zaremba) [inline x/leverage/keeper/iter.go:178]: ditto -> err check

**Review by robert-zaremba — COMMENTED** [review]
I'm not sure if this is a right solution.
I'm worried about  _borrowed value will skip borrowed tokens with unknown oracle prices, treating them as zero value_ -- this can prevent liquidations, which normally should happen.

Also, 
- [x] proper error checks are missing.
- [ ] let's add a follow up task to test new edge cases
- [ ] see , document or make a task to solve panic in case of missing collateral price when we try to liquidate a position.
- [ ] probably, this will also block uncollaterize operations.

**Review by robert-zaremba — APPROVED** [review]
pre-approving to make the planned released sooner.
I've updated the previous review note: https://github.com/umee-network/umee/pull/1821#pullrequestreview-1298514938
Let's make sure we add all relevant tasks, and follow ups in the task board.

File: x/leverage/keeper/borrows.go (hunk 3 above)
Comment (toteki) [reply x/leverage/keeper/borrows.go:163]: I'll see if I can skip the three possible "missing price" errors:
- `decmath.Median`'s `ErrEmptyList` (from historic price mode)
- `leverage.ErrNoHistoricMedians` (same, but leverage side)
- `oracle.ErrUnknownDenom` (regular price mode)

**Comment by toteki** [comment]
> see , document or make a task to solve panic in case of missing collateral price when we try to liquidate a position.

This is an intended error return - we can't declare a borrower unhealthy (liquidation eligible) without knowing the value of all their collateral.

I'll check off the rest on your comment as they are done

**Comment by robert-zaremba** [comment]
> Decollateralize and withdraw both use assertBorrowerHealth, which requires the value of all borrows is known, and that the borrow limit derived from their collateral with known prices (some might be missing) is sufficient to cover the borrows.

We should enable a user to decollaterize the asset with zero price.
